### PR TITLE
[ApiBundle] Validate locale field on translation object

### PR DESCRIPTION
**Pull request by Wojdylak**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | N/A
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        $context[self::ALREADY_CALLED] = true;
+        $data = (array) $data;
+
+        if (array_key_exists('translations', $data)) {
+            $newKeys = array_map(static fn (array $translation) => $translation['locale'] ?? '', $data['translations']);
+            $data['translations'] = array_combine($newKeys, $data['translations']);
+        }
+
+        return $this->denormalizer->denormalize($data, $type, $format, $context);
+    }
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php]:
```suggestion
	/**
	* @param array<string, array{ translations: array<mixed> }> // we might improve it
	*/
    public function denormalize(mixed $data, string $type, string $format = null, array $context = [])
    {
        $context[self::ALREADY_CALLED] = true;
        //$data = (array) $data; <-- you're checking in `supportsDenormalization` whether the $data is an array or not.

        if (array_key_exists('translations', $data)) {
            $newKeys = array_map(static fn (array $translation) => $translation['locale'] ?? '', $data['translations']);
            $data['translations'] = array_combine($newKeys, $data['translations']);
        }

        return $this->denormalizer->denormalize($data, $type, $format, $context);
    }
```

File: src/Sylius/Bundle/ProductBundle/Resources/translations/validators.en.yml [hunk 2]
```diff
@@ -40,13 +44,17 @@ sylius:
             max_length: Option name must not be longer than 1 character.|Option name must not be longer than {{ limit }} characters.
         values:
             min_count: Please add at least {{ limit }} option value.|Please add at least {{ limit }} option values.
+        locale:
+            not_blank: Please enter the locale.
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ProductBundle/Resources/translations/validators.en.yml]: Maybe we can create some more generic translation key. We can create one translation key like this per bundle.

File: src/Sylius/Bundle/ApiBundle/OpenApi/Documentation/TranslatableLocaleKeyDocumentationModifier.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        $schemas = $components->getSchemas();
+
+        $actions = array_filter($schemas->getArrayCopy(), function ($key) {
+            return preg_match('/\.(create|update)$/', $key);
+        }, \ARRAY_FILTER_USE_KEY);
+
+        foreach ($actions as $key => $action) {
+            if (null === $properties = $action['properties'] ?? null) {
+                continue;
+            }
```
Comment (jakubtobiasz) [inline src/Sylius/Bundle/ApiBundle/OpenApi/Documentation/TranslatableLocaleKeyDocumentationModifier.php]:
```suggestion
			$properties = $action['properties'] ?? null;
            if (null === $properties) {
                continue;
            }
```

IMO the current version is less readable.

File: src/Sylius/Bundle/ApiBundle/OpenApi/Documentation/TranslatableLocaleKeyDocumentationModifier.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+            $translations = $properties['translations'] ?? null;
+            if (null === $translations) {
+                continue;
+            }
+
+            $example = $translations['example'] ?? null;
+            if (null === $example) {
+                continue;
+            }
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ApiBundle/OpenApi/Documentation/TranslatableLocaleKeyDocumentationModifier.php]:
```suggestion
            $example = $action['properties']['translations']['example'] ?? null;
            if (null === $example) {
                continue;
            }
```

File: tests/Api/Admin/ProductsTest.php [hunk 5]
```diff
@@ -147,6 +145,38 @@ public function it_does_not_create_a_product_without_required_data(): void
         );
     }
 
+    /** @test */
+    public function it_does_not_create_a_product_without_translation_locale(): void
```
Comment (Rafikooo) [inline tests/Api/Admin/ProductsTest.php:149]:
```suggestion
    public function it_does_not_create_a_product_with_invalid_translation_locale(): void
```

**Review by Rafikooo — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+use Symfony\Component\Serializer\Normalizer\ContextAwareDenormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\DenormalizerAwareTrait;
+
+/** @experimental */
+final class TranslatableLocaleKeyDenormalizer implements ContextAwareDenormalizerInterface, DenormalizerAwareInterface
+{
+    use DenormalizerAwareTrait;
+
+    private const ALREADY_CALLED = 'sylius_translatable_locale_key_denormalizer_already_called';
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php:26]: Will this denormalizer be called on relations, if one translatable has another, like product and main taxon?

Is it necessary for this denormalizer to be executed multiple times?

File: src/Sylius/Bundle/AttributeBundle/Resources/config/validation/AttributeTranslation.xml [hunk 7]
```diff
@@ -26,5 +26,15 @@
                 <option name="groups">sylius</option>
             </constraint>
         </property>
+        <property name="locale">
```
Comment (diimpp) [inline src/Sylius/Bundle/AttributeBundle/Resources/config/validation/AttributeTranslation.xml:29]: Do we need to check, that locale is exists in database? Or is it checked somewhere else?

File: tests/Api/Admin/ProductsTest.php [hunk 8]
```diff
@@ -147,6 +145,38 @@ public function it_does_not_create_a_product_without_required_data(): void
...
+                        'slug' => 'kubek',
+                        'name' => 'Kubek',
+                    ],
+                ],
+            ], JSON_THROW_ON_ERROR),
+        );
+
+        $this->assertResponse(
+            $this->client->getResponse(),
+            'admin/product/post_product_without_translation_locale',
```
Comment (GSadee) [inline tests/Api/Admin/ProductsTest.php:175]:
```suggestion
            'admin/product/post_product_with_invalid_translation_locale',
```

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Karol! :tada:

File: src/Sylius/Bundle/AttributeBundle/Resources/config/validation/AttributeTranslation.xml (hunk 7 above)
Comment (GSadee) [reply src/Sylius/Bundle/AttributeBundle/Resources/config/validation/AttributeTranslation.xml:29]: If I am not wrong, this is not checked anywhere, and probably should be, we will add it in a separate PR

File: src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php (hunk 6 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/ApiBundle/Serializer/TranslatableLocaleKeyDenormalizer.php:26]: #15535

**Pull request merged**